<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Function to calculate correlation between prioritised genes...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for xPierCor"><tr><td>xPierCor</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Function to calculate correlation between prioritised genes and user-defined external data</h2>

<h3>Description</h3>

<p><code>xPierCor</code> is supposed to calculate correlation between
prioritised genes and user-defined external data.
</p>


<h3>Usage</h3>

<pre>
xPierCor(pNode, data, method = c("pearson", "spearman"),
pvalue.type = c("nominal", "empirical"), seed = 825, nperm = 2000,
plot = FALSE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>pNode</code></td>
<td>
<p>an object of class &quot;pNode&quot; (or &quot;sTarget&quot; or &quot;dTarget&quot;).
Alternatively, it can be a data frame with two columns ('name' and
'priority')</p>
</td></tr>
<tr valign="top"><td><code>data</code></td>
<td>
<p>a named input vector containing numeric data for genes
(gene symbols)</p>
</td></tr>
<tr valign="top"><td><code>method</code></td>
<td>
<p>the method used to calcualte correlation. It can be
'pearson' for Pearson's correlation or 'spearman' for Spearman rank
correlation</p>
</td></tr>
<tr valign="top"><td><code>pvalue.type</code></td>
<td>
<p>the type of the p-value calcualted. It can be
'nominal' for nominal p-value or 'empirical' for empirical p-value</p>
</td></tr>
<tr valign="top"><td><code>seed</code></td>
<td>
<p>an integer specifying the seed</p>
</td></tr>
<tr valign="top"><td><code>nperm</code></td>
<td>
<p>the number of random permutations</p>
</td></tr>
<tr valign="top"><td><code>plot</code></td>
<td>
<p>logical to indicate whether scatter plot is drawn</p>
</td></tr>
</table>


<h3>Value</h3>

<p>an object of class &quot;ggplot&quot; with a component 'res' (a data frame with
two columns 'cor' and 'pval') if the scatter plot is drawn. Otherwise
it is a data frame with two columns 'cor' and 'pval'.
</p>


<h3>Note</h3>

<p>none
</p>


<h3>See Also</h3>

<p><code>xPierCor</code>
</p>


<h3>Examples</h3>

<pre>
## Not run: 
# Load the library
library(Pi)

## End(Not run)

RData.location &lt;- "http://galahad.well.ox.ac.uk/bigdata_dev"
## Not run: 
# a) provide the seed nodes/genes with the weight info
## load ImmunoBase
ImmunoBase &lt;- xRDataLoader(RData.customised='ImmunoBase',
RData.location=RData.location)
## get genes within 500kb away from AS GWAS lead SNPs
seeds.genes &lt;- ImmunoBase$AS$genes_variants
## seeds weighted according to distance away from lead SNPs
data &lt;- 1- seeds.genes/500000

# b) perform priority analysis
pNode &lt;- xPierGenes(data=data, network="PCommonsDN_medium",restart=0.7,
RData.location=RData.location)

# c) do correlation
data &lt;- pNode$priority$priority[1:100]
name(data) &lt;- pNode$priority$name[1:100]
res &lt;- xPierCor(pNode, data, method="pearson", pvalue.type="empirical",
nperm=2000, plot=TRUE)

## End(Not run)
</pre>


</body></html>
