<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Function to prioritise pathways based on GSEA analysis of...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for xPierGSEA"><tr><td>xPierGSEA</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Function to prioritise pathways based on GSEA analysis of prioritised genes</h2>

<h3>Description</h3>

<p><code>xPierGSEA</code> is supposed to prioritise pathways given prioritised
genes and the ontology in query. It is done via gene set enrichment
analysis (GSEA). It returns an object of class &quot;eGSEA&quot;.
</p>


<h3>Usage</h3>

<pre>
xPierGSEA(pNode, priority.top = NULL, ontology = c("GOBP", "GOMF",
"GOCC",
"PS", "PS2", "SF", "Pfam", "DO", "HPPA", "HPMI", "HPCM", "HPMA", "MP",
"EF",
"MsigdbH", "MsigdbC1", "MsigdbC2CGP", "MsigdbC2CPall", "MsigdbC2CP",
"MsigdbC2KEGG", "MsigdbC2REACTOME", "MsigdbC2BIOCARTA", "MsigdbC3TFT",
"MsigdbC3MIR", "MsigdbC4CGN", "MsigdbC4CM", "MsigdbC5BP", "MsigdbC5MF",
"MsigdbC5CC", "MsigdbC6", "MsigdbC7", "DGIdb", "GTExV4", "GTExV6p",
"GTExV7",
"CreedsDisease", "CreedsDiseaseUP", "CreedsDiseaseDN", "CreedsDrug",
"CreedsDrugUP", "CreedsDrugDN", "CreedsGene", "CreedsGeneUP",
"CreedsGeneDN", "KEGG", "KEGGmetabolism", "KEGGgenetic",
"KEGGenvironmental",
"KEGGcellular", "KEGGorganismal", "KEGGdisease"),
customised.genesets = NULL, size.range = c(10, 500),
p.adjust.method = c("BH", "BY", "bonferroni", "holm", "hochberg",
"hommel"),
path.mode = c("all_paths", "shortest_paths", "all_shortest_paths"),
weight = 1, seed = 825, nperm = 2000, fast = TRUE, verbose = TRUE,
silent = FALSE, RData.location =
"http://galahad.well.ox.ac.uk/bigdata")
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>pNode</code></td>
<td>
<p>an object of class &quot;pNode&quot; (or &quot;sTarget&quot; or &quot;dTarget&quot;).
Alternatively, it can be a data frame with two columns ('priority' and
'rank')</p>
</td></tr>
<tr valign="top"><td><code>priority.top</code></td>
<td>
<p>the number of the top targets used for GSEA. By
default, it is NULL meaning all targets are used</p>
</td></tr>
<tr valign="top"><td><code>ontology</code></td>
<td>
<p>the ontology supported currently. It can be &quot;GOBP&quot; for
Gene Ontology Biological Process, &quot;GOMF&quot; for Gene Ontology Molecular
Function, &quot;GOCC&quot; for Gene Ontology Cellular Component, &quot;PS&quot; for
phylostratific age information, &quot;PS2&quot; for the collapsed PS version
(inferred ancestors being collapsed into one with the known taxonomy
information), &quot;SF&quot; for SCOP domain superfamilies, &quot;Pfam&quot; for Pfam
domain families, &quot;DO&quot; for Disease Ontology, &quot;HPPA&quot; for Human Phenotype
Phenotypic Abnormality, &quot;HPMI&quot; for Human Phenotype Mode of Inheritance,
&quot;HPCM&quot; for Human Phenotype Clinical Modifier, &quot;HPMA&quot; for Human
Phenotype Mortality Aging, &quot;MP&quot; for Mammalian Phenotype, &quot;EF&quot; for
Experimental Factor Ontology (used to annotate GWAS Catalog genes),
Drug-Gene Interaction database (&quot;DGIdb&quot;) for drugable categories,
tissue-specific eQTL-containing genes from GTEx (&quot;GTExV4&quot;, &quot;GTExV6p&quot;
and &quot;GTExV7&quot;), crowd extracted expression of differential signatures
from CREEDS (&quot;CreedsDisease&quot;, &quot;CreedsDiseaseUP&quot;, &quot;CreedsDiseaseDN&quot;,
&quot;CreedsDrug&quot;, &quot;CreedsDrugUP&quot;, &quot;CreedsDrugDN&quot;, &quot;CreedsGene&quot;,
&quot;CreedsGeneUP&quot; and &quot;CreedsGeneDN&quot;), KEGG pathways (including 'KEGG' for
all, 'KEGGmetabolism' for 'Metabolism' pathways, 'KEGGgenetic' for
'Genetic Information Processing' pathways, 'KEGGenvironmental' for
'Environmental Information Processing' pathways, 'KEGGcellular' for
'Cellular Processes' pathways, 'KEGGorganismal' for 'Organismal
Systems' pathways, and 'KEGGdisease' for 'Human Diseases' pathways),
and the molecular signatures database (Msigdb, including &quot;MsigdbH&quot;,
&quot;MsigdbC1&quot;, &quot;MsigdbC2CGP&quot;, &quot;MsigdbC2CPall&quot;, &quot;MsigdbC2CP&quot;,
&quot;MsigdbC2KEGG&quot;, &quot;MsigdbC2REACTOME&quot;, &quot;MsigdbC2BIOCARTA&quot;, &quot;MsigdbC3TFT&quot;,
&quot;MsigdbC3MIR&quot;, &quot;MsigdbC4CGN&quot;, &quot;MsigdbC4CM&quot;, &quot;MsigdbC5BP&quot;, &quot;MsigdbC5MF&quot;,
&quot;MsigdbC5CC&quot;, &quot;MsigdbC6&quot;, &quot;MsigdbC7&quot;)</p>
</td></tr>
<tr valign="top"><td><code>customised.genesets</code></td>
<td>
<p>a list each containing gene symbols. By
default, it is NULL. If the list provided, it will overtake the
previous parameter &quot;ontology&quot;</p>
</td></tr>
<tr valign="top"><td><code>size.range</code></td>
<td>
<p>the minimum and maximum size of members of each term
in consideration. By default, it sets to a minimum of 10 but no more
than 500</p>
</td></tr>
<tr valign="top"><td><code>p.adjust.method</code></td>
<td>
<p>the method used to adjust p-values. It can be
one of &quot;BH&quot;, &quot;BY&quot;, &quot;bonferroni&quot;, &quot;holm&quot;, &quot;hochberg&quot; and &quot;hommel&quot;. The
first two methods &quot;BH&quot; (widely used) and &quot;BY&quot; control the false
discovery rate (FDR: the expected proportion of false discoveries
amongst the rejected hypotheses); the last four methods &quot;bonferroni&quot;,
&quot;holm&quot;, &quot;hochberg&quot; and &quot;hommel&quot; are designed to give strong control of
the family-wise error rate (FWER). Notes: FDR is a less stringent
condition than FWER</p>
</td></tr>
<tr valign="top"><td><code>path.mode</code></td>
<td>
<p>the mode of paths induced by vertices/nodes with input
annotation data. It can be &quot;all_paths&quot; for all possible paths to the
root, &quot;shortest_paths&quot; for only one path to the root (for each node in
query), &quot;all_shortest_paths&quot; for all shortest paths to the root (i.e.
for each node, find all shortest paths with the equal lengths)</p>
</td></tr>
<tr valign="top"><td><code>weight</code></td>
<td>
<p>an integer specifying score weight. It can be &quot;0&quot; for
unweighted (an equivalent to Kolmogorov-Smirnov, only considering the
rank), &quot;1&quot; for weighted by input gene score (by default), and &quot;2&quot; for
over-weighted, and so on</p>
</td></tr>
<tr valign="top"><td><code>seed</code></td>
<td>
<p>an integer specifying the seed</p>
</td></tr>
<tr valign="top"><td><code>nperm</code></td>
<td>
<p>the number of random permutations. For each permutation,
gene-score associations will be permutated so that permutation of
gene-term associations is realised</p>
</td></tr>
<tr valign="top"><td><code>fast</code></td>
<td>
<p>logical to indicate whether to fast calculate GSEA
resulting. By default, it sets to true, but not necessarily does so. It
will depend on whether the package &quot;fgsea&quot; has been installed. It can
be installed via: <code>source("http://bioconductor.org/biocLite.R");
biocLite(c("fgsea"))</code>. If not yet installed, this option will be
disabled</p>
</td></tr>
<tr valign="top"><td><code>verbose</code></td>
<td>
<p>logical to indicate whether the messages will be
displayed in the screen. By default, it sets to true</p>
</td></tr>
<tr valign="top"><td><code>silent</code></td>
<td>
<p>logical to indicate whether the messages will be silent
completely. By default, it sets to false. If true, verbose will be
forced to be false</p>
</td></tr>
<tr valign="top"><td><code>RData.location</code></td>
<td>
<p>the characters to tell the location of built-in
RData files. See <code>xRDataLoader</code> for details</p>
</td></tr>
</table>


<h3>Value</h3>

<p>an object of class &quot;eGSEA&quot;, a list with following components:
</p>

<ul>
<li><p><code>df_summary</code>: a data frame of nTerm x 9 containing gene set
enrichment analysis result, where nTerm is the number of
terms/genesets, and the 9 columns are &quot;setID&quot; (i.e. &quot;Term ID&quot;), &quot;name&quot;
(i.e. &quot;Term Name&quot;), &quot;nAnno&quot; (i.e. number in members annotated by a
term), &quot;nLead&quot; (i.e. number in members as leading genes), &quot;peak&quot; (i.e.
the rank at peak), &quot;total&quot; (i.e. the total number of genes analysed),
&quot;es&quot; (i.e. enrichment score), &quot;nes&quot; (i.e. normalised enrichment score;
enrichment score but after being normalised by gene set size), &quot;pvalue&quot;
(i.e. nominal p value), &quot;adjp&quot; (i.e. adjusted p value; p value but
after being adjusted for multiple comparisons), &quot;distance&quot; (i.e. term
distance or metadata)
</p>
</li>
<li><p><code>leading</code>: a list of gene sets, each storing leading gene
info (i.e. the named vector with names for gene symbols and elements
for priority rank). Always, gene sets are identified by &quot;setID&quot;
</p>
</li>
<li><p><code>full</code>: a list of gene sets, each storing full info on gene
set enrichment analysis result (i.e. a data frame of nGene x 6, where
nGene is the number of genes, and the 6 columns are &quot;GeneID&quot;, &quot;Rank&quot;
for priority rank, &quot;Score&quot; for priority score, &quot;RES&quot; for running
enrichment score, &quot;Hits&quot; for gene set hits info with 1 for gene hit, 2
for leading gene hit, 3 for the point defining leading genes, 0 for no
hit), and &quot;Symbol&quot; for gene symbols. Always, gene sets are identified
by &quot;setID&quot;
</p>
</li>
<li><p><code>cross</code>: a matrix of nTerm X nTerm, with an on-diagnal cell
for the leading genes observed in an individaul term, and off-diagnal
cell for the overlapped leading genes shared between two terms
</p>
</li></ul>



<h3>Note</h3>

<p>none
</p>


<h3>See Also</h3>

<p><code>xGSEAbarplot</code>, <code>xGSEAdotplot</code>
</p>


<h3>Examples</h3>

<pre>
## Not run: 
# Load the library
library(Pi)

## End(Not run)

RData.location &lt;- "http://galahad.well.ox.ac.uk/bigdata"
## Not run: 
# a) provide the seed nodes/genes with the weight info
## load ImmunoBase
ImmunoBase &lt;- xRDataLoader(RData.customised='ImmunoBase',
RData.location=RData.location)
## get genes within 500kb away from AS GWAS lead SNPs
seeds.genes &lt;- ImmunoBase$AS$genes_variants
## seeds weighted according to distance away from lead SNPs
data &lt;- 1- seeds.genes/500000

# b) perform priority analysis
pNode &lt;- xPierGenes(data=data, network="PCommonsDN_medium",restart=0.7,
RData.location=RData.location)

# c) do pathway-level priority using GSEA
eGSEA &lt;- xPierGSEA(pNode=pNode, ontology="DGIdb", nperm=2000,
RData.location=RData.location)
bp &lt;- xGSEAbarplot(eGSEA, top_num="auto", displayBy="nes")
gp &lt;- xGSEAdotplot(eGSEA, top=1)

## End(Not run)
</pre>


</body></html>
