<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Function to define HiC genes given a list of SNPs</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for xSNP2cGenes"><tr><td>xSNP2cGenes</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Function to define HiC genes given a list of SNPs</h2>

<h3>Description</h3>

<p><code>xSNP2cGenes</code> is supposed to define HiC genes given a list of
SNPs. The HiC weight is calcualted as Cumulative Distribution Function
of HiC interaction scores.
</p>


<h3>Usage</h3>

<pre>
xSNP2cGenes(data, entity = c("SNP", "chr:start-end", "data.frame",
"bed",
"GRanges"), include.HiC = c(NA, "Monocytes", "Macrophages_M0",
"Macrophages_M1", "Macrophages_M2", "Neutrophils", "Megakaryocytes",
"Endothelial_precursors", "Erythroblasts", "Fetal_thymus",
"Naive_CD4_T_cells", "Total_CD4_T_cells",
"Activated_total_CD4_T_cells",
"Nonactivated_total_CD4_T_cells", "Naive_CD8_T_cells",
"Total_CD8_T_cells",
"Naive_B_cells", "Total_B_cells"), GR.SNP = c("dbSNP_GWAS",
"dbSNP_Common"),
cdf.function = c("empirical", "exponential"), plot = FALSE,
verbose = TRUE, RData.location =
"http://galahad.well.ox.ac.uk/bigdata")
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>data</code></td>
<td>
<p>NULL or a input vector containing SNPs. If NULL, all SNPs
will be considered. If a input vector containing SNPs, SNPs should be
provided as dbSNP ID (ie starting with rs) or in the format of
'chrN:xxx', where N is either 1-22 or X, xxx is number; for example,
'chr16:28525386'. Alternatively, it can be other formats/entities (see
the next parameter 'entity')</p>
</td></tr>
<tr valign="top"><td><code>entity</code></td>
<td>
<p>the data entity. By default, it is &quot;SNP&quot;. For general
use, it can also be one of &quot;chr:start-end&quot;, &quot;data.frame&quot;, &quot;bed&quot; or
&quot;GRanges&quot;</p>
</td></tr>
<tr valign="top"><td><code>include.HiC</code></td>
<td>
<p>genes linked to input SNPs are also included. By
default, it is 'NA' to disable this option. Genes linked to input SNPs
are based on Promoter Capture HiC (PCHic), including 17 primary blood
cell types:
&quot;Monocytes&quot;,&quot;Macrophages_M0&quot;,&quot;Macrophages_M1&quot;,&quot;Macrophages_M2&quot;,&quot;Neutrophils&quot;,&quot;Megakaryocytes&quot;,&quot;Endothelial_precursors&quot;,&quot;Erythroblasts&quot;,&quot;Fetal_thymus&quot;,&quot;Naive_CD4_T_cells&quot;,&quot;Total_CD4_T_cells&quot;,&quot;Activated_total_CD4_T_cells&quot;,&quot;Nonactivated_total_CD4_T_cells&quot;,&quot;Naive_CD8_T_cells&quot;,&quot;Total_CD8_T_cells&quot;,&quot;Naive_B_cells&quot;,&quot;Total_B_cells&quot;.</p>
</td></tr>
<tr valign="top"><td><code>GR.SNP</code></td>
<td>
<p>the genomic regions of SNPs. By default, it is
'dbSNP_GWAS', that is, SNPs from dbSNP (version 146) restricted to GWAS
SNPs and their LD SNPs (hg19). It can be 'dbSNP_Common', that is,
Common SNPs from dbSNP (version 146) plus GWAS SNPs and their LD SNPs
(hg19). Alternatively, the user can specify the customised input. To do
so, first save your RData file (containing an GR object) into your
local computer, and make sure the GR object content names refer to
dbSNP IDs. Then, tell &quot;GR.SNP&quot; with your RData file name (with or
without extension), plus specify your file RData path in
&quot;RData.location&quot;. Note: you can also load your customised GR object
directly</p>
</td></tr>
<tr valign="top"><td><code>cdf.function</code></td>
<td>
<p>a character specifying a Cumulative Distribution
Function (cdf). It can be one of 'exponential' based on exponential
cdf, 'empirical' for empirical cdf</p>
</td></tr>
<tr valign="top"><td><code>plot</code></td>
<td>
<p>logical to indicate whether the histogram plot (plus
density or CDF plot) should be drawn. By default, it sets to false for
no plotting</p>
</td></tr>
<tr valign="top"><td><code>verbose</code></td>
<td>
<p>logical to indicate whether the messages will be
displayed in the screen. By default, it sets to true for display</p>
</td></tr>
<tr valign="top"><td><code>RData.location</code></td>
<td>
<p>the characters to tell the location of built-in
RData files. See <code>xRDataLoader</code> for details</p>
</td></tr>
</table>


<h3>Value</h3>

<p>a data frame with following columns:
</p>

<ul>
<li><p><code>Gene</code>: SNP-interacting genes caputured by HiC
</p>
</li>
<li><p><code>SNP</code>: SNPs
</p>
</li>
<li><p><code>Sig</code>: the interaction score (the higher stronger)
</p>
</li>
<li><p><code>Weight</code>: the HiC weight
</p>
</li></ul>



<h3>Note</h3>

<p>None
</p>


<h3>See Also</h3>

<p><code>xSNPhic</code>
</p>


<h3>Examples</h3>

<pre>
## Not run: 
# Load the library
library(Pi)

## End(Not run)

RData.location &lt;- "http://galahad.well.ox.ac.uk/bigdata_dev"
# a) provide the SNPs with the significance info
## get lead SNPs reported in AS GWAS and their significance info (p-values)
#data.file &lt;- "http://galahad.well.ox.ac.uk/bigdata/AS.txt"
#AS &lt;- read.delim(data.file, header=TRUE, stringsAsFactors=FALSE)
ImmunoBase &lt;- xRDataLoader(RData.customised='ImmunoBase',
RData.location=RData.location)
data &lt;- names(ImmunoBase$AS$variants)

## Not run: 
# b) define HiC genes
df_cGenes &lt;- xSNP2cGenes(data, include.HiC="Monocytes",
RData.location=RData.location)

## End(Not run)
</pre>


</body></html>
